File: app/src/main/java/com/osrsd/cache/util/ByteBufferExt.java
package com.osrsd.cache.util;

import java.nio.ByteBuffer;

public final class ByteBufferExt {

    public static int getMedium(ByteBuffer buffer) {
        return ((buffer.get() & 0xff) << 16) | ((buffer.get() & 0xff) << 8) | (buffer.get() & 0xff);
    }

    public static String getString(ByteBuffer buffer) {
        StringBuilder builder = new StringBuilder();
        int b;
        while ((b = (buffer.get() & 0xff)) != 0) {
            builder.append((char) b);
        }
        return builder.toString();
    }

    public static String getStringOrNull(ByteBuffer buffer) {
        int peek = buffer.get(buffer.position()) & 0xff;
        if (peek != 0) {
            return getString(buffer);
        } else {
            buffer.get(); // discard
            return null;
        }
    }

    public static int getShortSmart(ByteBuffer buffer) {
        int peek = buffer.get(buffer.position()) & 0xff;
        return peek < 128 ? (buffer.get() & 0xff) - 64 : (buffer.getShort() & 0xffff) - 0xc000;
    }

    public static int getUnsignedShortSmart(ByteBuffer buffer) {
        int peek = buffer.get(buffer.position()) & 0xff;
        return peek < 128 ? (buffer.get() & 0xff) : (buffer.getShort() & 0xffff) - 0x8000;
    }

}

File: app/src/main/java/com/osrsd/cache/util/ByteBufferExt.kt
package com.osrsd.cache.util

import java.nio.ByteBuffer

object ByteBufferExt {

    fun getMedium(buffer: ByteBuffer): Int {
        return buffer.get().toInt() and 0xff shl 16 or (buffer.get().toInt() and 0xff shl 8) or (buffer.get().toInt() and 0xff)
    }

    fun getString(buffer: ByteBuffer): String {
        val builder = StringBuilder()
        var b: Int
        while ((buffer.get().toInt() and 0xff).also { b = it } != 0) {
            builder.append(b.toChar())
        }
        return builder.toString()
    }

    fun getStringOrNull(buffer: ByteBuffer): String {
        val peek: Int = buffer[buffer.position()].toInt() and 0xff
        return if (peek != 0) {
            getString(buffer)
        } else {
            buffer.get() // discard
            "null"
        }
    }

    fun getShortSmart(buffer: ByteBuffer): Int {
        val peek: Int = buffer[buffer.position()].toInt() and 0xff
        return if (peek < 128) (buffer.get().toInt() and 0xff) - 64 else (buffer.short.toInt() and 0xffff) - 0xc000
    }

    fun getUnsignedShortSmart(buffer: ByteBuffer): Int {
        val peek: Int = buffer[buffer.position()].toInt() and 0xff
        return if (peek < 128) buffer.get().toInt() and 0xff else (buffer.short.toInt() and 0xffff) - 0x8000
    }

}